File: plugins/durable-threads/skills/durable-threads/scripts/pro5.mjs
#!/usr/bin/env node
/** Local Pro-allowance diagnostics. No model turns, config writes, or quota resets. */
import { spawn } from 'node:child_process';
import { createReadStream, readFileSync, statSync } from 'node:fs';
import { createHash } from 'node:crypto';
import { pathToFileURL } from 'node:url';

const record = (v) => v !== null && typeof v === 'object' && !Array.isArray(v);
const body = (v) => record(v?.result) ? v.result : v;
const count = (v) => Number.isSafeInteger(v) && v >= 0;
const finite = (v) => typeof v === 'number' && Number.isFinite(v);
const identifier = (v) => typeof v === 'string' && /^[a-zA-Z0-9_.:/-]{1,128}$/.test(v);
const hash = (v) => createHash('sha256').update(v).digest('hex');
const META_METHODS = new Set(['initialize', 'account/read', 'account/rateLimits/read', 'model/list', 'experimentalFeature/list']);

/** A Pro plan label does not establish the 5x/20x entitlement. Missing != unlimited. */
export function quotaStatus(payload, nowSeconds, { limitId = 'codex', reservePercent = 10 } = {}) {
  if (!finite(nowSeconds) || nowSeconds < 0 || !finite(reservePercent) || reservePercent < 0 || reservePercent >= 100) {
    throw new Error('Invalid quota policy');
  }
  const p = body(payload);
  const snapshot = p?.rateLimitsByLimitId?.[limitId] ??
    (p?.rateLimits?.limitId === limitId ? p.rateLimits : null);
  if (!record(snapshot)) return { status: 'unknown', limitId, windows: [], reason: 'No matching allowance bucket' };
  const windows = [];
  let missing = false;
  for (const name of ['primary', 'secondary']) {
    const w = snapshot[name];
    if (w == null) continue;
    if (!record(w) || !finite(w.usedPercent) || w.usedPercent < 0 || w.usedPercent > 100 ||
        !count(w.resetsAt) || !count(w.windowDurationMins) || w.windowDurationMins === 0) {
      missing = true;
      continue;
    }
    windows.push({ name, usedPercent: w.usedPercent, remainingPercent: 100 - w.usedPercent,
      durationMinutes: w.windowDurationMins, resetsAt: w.resetsAt });
  }
  // A past reset is a reason to refresh, never permission to assume a fresh allowance.
  if (!windows.length || missing || windows.some(w => w.resetsAt <= nowSeconds)) {
    return { status: 'unknown', limitId, windows, reason: 'Missing, invalid, or expired window; refresh /status' };
  }
  if (snapshot.rateLimitReachedType != null || snapshot.spendControlReached === true ||
      windows.some(w => w.remainingPercent <= 0)) {
    return { status: 'exhausted', limitId, windows, reason: 'Stop new model work; retain a handoff' };
  }
  const remainingPercent = Math.min(...windows.map(w => w.remainingPercent));
  return { status: remainingPercent <= reservePercent ? 'reserve' : 'available', limitId,
    windows, remainingPercent, reason: 'Observed bucket only; not a token-to-quota estimate' };
}

/** Pure advisory governor. The plugin cannot intercept Codex inference or enforce account limits. */
export function advise(state, nowSeconds) {
  if (!record(state) || !finite(nowSeconds)) throw new Error('Expected an explicit state object and time');
  for (const key of ['cancelled', 'reviewRequired', 'reviewPassed', 'candidateComplete', 'checksPassed', 'scopeVerified', 'humanApproved', 'conceptualFailure', 'unresolvedCritical']) {
    if (key in state && typeof state[key] !== 'boolean') throw new Error(`Invalid flag: ${key}`);
  }
  const reply = (action, reason, effort = state.currentEffort ?? null) => ({
    action, reason, effort, advisoryOnly: true, scope: 'next-safe-boundary', liveControl: false,
  });
  if (state.cancelled === true) return reply('stop', 'Cancellation remains authoritative');
  if (state.authMode !== 'chatgpt') return reply('stop', 'Verify ChatGPT sign-in; do not fall back to paid API billing');
  if (state.writerState !== 'known') return reply('stop', 'Reconcile writer state before retrying');
  if (!['R0', 'R1', 'R2', 'R3', 'R4'].includes(state.risk)) throw new Error('Invalid risk class');
  const reviewRequired = state.reviewRequired === true || ['R3', 'R4'].includes(state.risk);
  if (state.candidateComplete === true && state.checksPassed === true && state.scopeVerified === true) {
    return reviewRequired && state.reviewPassed !== true
      ? reply('review', 'Independent review is still required')
      : reply('complete', 'Acceptance passed; stop without optional repeat reviews');
  }
  for (const key of ['generations', 'noProgressGenerations', 'repeatedFailures', 'elapsedSeconds', 'stepsSinceChange', 'resolvedCheckpoints']) {
    if (!count(state[key])) throw new Error(`Missing or invalid counter: ${key}`);
  }
  // These are DT experiment defaults, not the subscription's rate limits.
  if (state.generations >= 60 || state.elapsedSeconds >= 1800 ||
      state.noProgressGenerations >= 4 || state.repeatedFailures >= 2) {
    return reply('pause', 'Bounded run/circuit-breaker reached; diagnose or revise the contract');
  }
  if (!finite(state.observedAt) || state.observedAt > nowSeconds || nowSeconds - state.observedAt > 300) {
    return reply('refresh_quota', 'Allowance observation is missing or stale');
  }
  const quota = quotaStatus(state.rateLimits, nowSeconds, { limitId: state.limitId ?? 'codex' });
  if (quota.status === 'unknown') return reply('refresh_quota', quota.reason);
  if (quota.status === 'exhausted' || quota.status === 'reserve') return reply('pause', 'Preserve quota and required verification; do not silently weaken acceptance');
  if (state.risk === 'R4' && state.humanApproved !== true) return reply('review', 'Systemic work needs approval before execution');
  if (state.phase === 'waiting') return reply('wait', 'Use the existing native wait/event mechanism, not another inference');
  if (state.candidateComplete === true) return reply('verify', 'A completion claim or human approval cannot replace passing checks and scope evidence');
  if (!['read', 'implement', 'diagnose', 'review'].includes(state.phase)) throw new Error('Invalid work phase');
  const levels = ['low', 'medium', 'high', 'xhigh', 'max'];
  const supported = state.supportedEfforts;
  if (!Array.isArray(supported) || !supported.length || !supported.every(x => typeof x === 'string') ||
      !supported.includes(state.currentEffort) || !levels.includes(state.currentEffort)) {
    return reply('inspect_capabilities', 'Use live advertised effort levels; never invent a universal effort ladder');
  }
  let desired = state.phase === 'diagnose' || state.phase === 'review' || reviewRequired ? 'medium' : 'low';
  if (state.conceptualFailure === true || state.unresolvedCritical === true) desired = 'high';
  const target = levels.find(x => levels.indexOf(x) >= levels.indexOf(desired) &&
    levels.indexOf(x) <= 2 && supported.includes(x));
  if (!target) return reply('inspect_capabilities', 'No supported effort within the automatic high-effort cap');
  const down = levels.indexOf(target) < levels.indexOf(state.currentEffort);
  if (down && (state.stepsSinceChange < 2 || state.resolvedCheckpoints < 2 || state.unresolvedCritical === true)) {
    return reply('keep', 'Downshift requires two resolved checkpoints and a two-step cooldown');
  }
  return reply(target === state.currentEffort ? 'keep' : 'set_effort',
    'Same explicitly selected model; validate quality after the next bounded work unit', target);
}

/** Read only explicitly selected rollouts. Ignore cumulative mirrors and nested compaction copies. */
export async function auditUsage(paths, { maxBytes = 256 * 1024 * 1024, maxLineBytes = 4 * 1024 * 1024 } = {}) {
  if (!Array.isArray(paths) || !paths.length || !paths.every(p => typeof p === 'string') ||
      !count(maxBytes) || maxBytes === 0 || !count(maxLineBytes) || maxLineBytes === 0) throw new Error('Invalid audit input');
  const seen = new Map();
  const totals = { inputTokens: 0, cachedInputTokens: 0, outputTokens: 0, reasoningOutputTokens: 0, cacheWriteInputTokens: 0 };
  const coverage = { lines: 0, ignored: 0, malformed: 0, oversized: 0, invalidUsage: 0,
    unidentified: 0, duplicates: 0, conflictingDuplicates: 0, missingOptional: 0, tailFragments: 0 };
  let responses = 0;
  const consume = (line) => {
    coverage.lines++;
    let row;
    try { row = JSON.parse(line); } catch { coverage.malformed++; return; }
    if (row?.type !== 'token_usage_record') { coverage.ignored++; return; }
    const p = row.payload;
    if (!record(p) || !identifier(p.response_id) || !identifier(p.thread_id)) { coverage.unidentified++; return; }
    const u = p.usage;
    if (!record(u) || ![u.input_tokens, u.cached_input_tokens, u.output_tokens].every(count) ||
        u.cached_input_tokens > u.input_tokens ||
        (u.reasoning_output_tokens != null && (!count(u.reasoning_output_tokens) || u.reasoning_output_tokens > u.output_tokens)) ||
        (u.cache_write_input_tokens != null && (!count(u.cache_write_input_tokens) || u.cache_write_input_tokens > u.input_tokens))) {
      coverage.invalidUsage++; return;
    }
    const values = [u.input_tokens, u.cached_input_tokens, u.output_tokens, u.reasoning_output_tokens ?? 0, u.cache_write_input_tokens ?? 0];
    const key = hash(`${p.thread_id}\0${p.response_id}`);
    const signature = values.join(',');
    if (seen.has(key)) {
      coverage.duplicates++;
      if (seen.get(key) !== signature) coverage.conflictingDuplicates++;
      return;
    }
    const keys = Object.keys(totals);
    if (keys.some((k, i) => !Number.isSafeInteger(totals[k] + values[i])) ||
        !Number.isSafeInteger(totals.inputTokens + values[0] + totals.outputTokens + values[2])) throw new Error('Usage total exceeds safe integer range');
    seen.set(key, signature);
    responses++;
    if (u.reasoning_output_tokens == null || u.cache_write_input_tokens == null) coverage.missingOptional++;
    keys.forEach((k, i) => { totals[k] += values[i]; });
  };
  let bytes = 0;
  for (const path of paths) {
    let pending = Buffer.alloc(0);
    let dropping = false;
    for await (const chunk of createReadStream(path, { highWaterMark: 64 * 1024 })) {
      bytes += chunk.length;
      if (bytes > maxBytes) throw new Error('Audit byte budget exceeded; select a smaller explicit sample');
      const joined = Buffer.concat([pending, chunk]);
      let start = 0;
      for (let i = 0; i < joined.length; i++) {
        if (joined[i] !== 10) continue;
        if (dropping) dropping = false;
        else if (i - start > maxLineBytes) coverage.oversized++;
        else if (i > start) consume(joined.subarray(start, i).toString('utf8'));
        start = i + 1;
      }
      pending = joined.subarray(start);
      if (pending.length > maxLineBytes) {
        if (!dropping) coverage.oversized++;
        dropping = true;
        pending = Buffer.alloc(0);
      }
    }
    if (pending.length || dropping) coverage.tailFragments++; // Active writers can leave an incomplete final line.
  }
  return { schemaVersion: 1, observedResponses: responses, ...totals,
    uncachedInputTokens: totals.inputTokens - totals.cachedInputTokens,
    inputPlusOutputTokens: totals.inputTokens + totals.outputTokens,
    cachedInputRatio: totals.inputTokens ? totals.cachedInputTokens / totals.inputTokens : null,
    coverage, billingEstimate: null, remainingAllowanceEstimate: null,
    caveat: 'Completed response records in selected files only. Reasoning is a subset of output. Warmups, compactions, children, and other products may be missing. No billing or quota inference.' };
}

/** Metadata-only stdio transport; method allowlist prevents accidental model execution. */
export async function probe({ command = 'codex', args = ['app-server'], timeoutMs = 10000 } = {}) {
  if (!count(timeoutMs) || timeoutMs < 100 || timeoutMs > 60000) throw new Error('Invalid probe timeout');
  const child = spawn(command, args, { stdio: ['pipe', 'pipe', 'pipe'], shell: false });
  child.stdout.setEncoding('utf8');
  child.stderr.resume(); // Never echo raw server errors/config/credentials.
  const pending = new Map();
  const errors = [];
  let seq = 0;
  let buffer = '';
  let fatal = false;
  const fail = () => {
    fatal = true;
    for (const p of pending.values()) { clearTimeout(p.timer); p.reject(new Error('Metadata transport failed')); }
    pending.clear();
  };
  child.on('error', fail);
  child.on('exit', fail);
  child.stdin.on('error', fail);
  child.stdout.on('data', chunk => {
    buffer += chunk.toString('utf8');
    if (Buffer.byteLength(buffer) > 8 * 1024 * 1024) { fail(); child.kill(); return; }
    let newline;
    while ((newline = buffer.indexOf('\n')) >= 0) {
      const line = buffer.slice(0, newline); buffer = buffer.slice(newline + 1);
      let message;
      try { message = JSON.parse(line); } catch { continue; }
      if (message.method && message.id !== undefined) {
        // Fail closed on unexpected server requests rather than approve tools or login.
        child.stdin.write(JSON.stringify({ id: message.id, error: { code: -32601, message: 'Metadata-only client' } }) + '\n');
        continue;
      }
      const p = pending.get(message.id);
      if (!p) continue;
      clearTimeout(p.timer); pending.delete(message.id);
      if (message.error) p.reject(new Error('Metadata method unavailable'));
      else p.resolve(message.result);
    }
  });
  const request = (method, params) => {
    if (!META_METHODS.has(method) || fatal) return Promise.reject(new Error('Metadata transport unavailable'));
    return new Promise((resolve, reject) => {
      const id = ++seq;
      const timer = setTimeout(() => { pending.delete(id); reject(new Error('Metadata timeout')); }, timeoutMs);
      pending.set(id, { resolve, reject, timer });
      child.stdin.write(JSON.stringify({ id, method, params }) + '\n');
    });
  };
  const safe = async (method, params) => {
    try { return await request(method, params); } catch { errors.push(method); return null; }
  };
  const page = async (method, extra) => {
    const data = []; const cursors = new Set(); let cursor;
    for (let i = 0; i < 20; i++) {
      const p = await safe(method, { ...extra, limit: 100, ...(cursor ? { cursor } : {}) });
      if (!Array.isArray(p?.data)) return { data, complete: false };
      data.push(...p.data);
      if (p.nextCursor == null) return { data, complete: true };
      if (typeof p.nextCursor !== 'string' || !p.nextCursor || cursors.has(p.nextCursor)) return { data, complete: false };
      cursor = p.nextCursor; cursors.add(cursor);
    }
    return { data, complete: false };
  };
  try {
    const initialized = await safe('initialize', { clientInfo: { name: 'durable_threads_pro5', version: '0.1.0' }, capabilities: { experimentalApi: false } });
    if (initialized == null) throw new Error('Codex metadata initialization failed');
    child.stdin.write(JSON.stringify({ method: 'initialized', params: {} }) + '\n');
    const account = await safe('account/read', { refreshToken: false });
    const rates = await safe('account/rateLimits/read', {});
    const models = await page('model/list', { includeHidden: false });
    const features = await page('experimentalFeature/list', {});
    const now = Math.floor(Date.now() / 1000);
    return {
      schemaVersion: 1, observedAt: now, requestedProfile: 'pro5-astra', planMultiplierVerified: false,
      authMode: account?.account?.type === 'chatgpt' ? 'chatgpt' : account?.account?.type === 'apiKey' ? 'apiKey' : 'unknown',
      planType: ['pro', 'plus', 'free', 'team', 'business', 'enterprise', 'edu'].includes(account?.account?.planType) ? account.account.planType : 'unknown',
      quota: quotaStatus(rates, now),
      quotaBuckets: Object.keys(rates?.rateLimitsByLimitId ?? {}).filter(identifier).map(limitId => quotaStatus(rates, now, { limitId })),
      modelCatalogComplete: models.complete,
      models: models.data.filter(m => identifier(m?.model ?? m?.id)).map(m => ({
        model: m.model ?? m.id,
        supportedEfforts: Array.isArray(m.supportedReasoningEfforts) ? m.supportedReasoningEfforts.map(e => e?.reasoningEffort).filter(identifier) : [],
        defaultEffort: identifier(m.defaultReasoningEffort) ? m.defaultReasoningEffort : null,
      })),
      featureCatalogComplete: features.complete,
      experimentalFeatures: features.data.filter(f => ['step_model_switching', 'reasoning_effort_override'].includes(f?.name)).map(f => ({ name: f.name, enabled: f.enabled === true })),
      errors: [...new Set(errors)], liveStepControlVerified: false,
    };
  } finally {
    fail(); child.stdin.end(); child.kill('SIGTERM');
    const timer = setTimeout(() => child.kill('SIGKILL'), 1000); timer.unref();
    if (child.exitCode === null && child.signalCode === null) await new Promise(resolve => {
      child.once('exit', resolve); child.once('error', resolve);
      const fallback = setTimeout(resolve, 1500); fallback.unref();
    });
    clearTimeout(timer);
  }
}

async function main(argv) {
  const [command, ...paths] = argv;
  if (command === 'probe' && paths.length === 0) return probe();
  if (command === 'audit' && paths.length) return auditUsage(paths);
  if (command === 'advise' && paths.length === 1) {
    if (statSync(paths[0]).size > 1024 * 1024) throw new Error('State file is too large');
    const raw = readFileSync(paths[0]);
    if (raw.length > 1024 * 1024) throw new Error('State file is too large');
    return advise(JSON.parse(raw), Math.floor(Date.now() / 1000));
  }
  throw new Error('Usage: node pro5.mjs probe | audit <explicit-rollout.jsonl> [...] | advise <state.json>');
}
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main(process.argv.slice(2)).then(value => console.log(JSON.stringify(value, null, 2))).catch(error => {
    // Never expose local paths, transcript excerpts, or server error bodies.
    console.error(error.message.startsWith('Usage:') ? error.message : 'Pro5 diagnostic failed. Check inputs, installed Codex, and local permissions. No settings were changed.');
    process.exitCode = 1;
  });
}

File: plugins/durable-threads/skills/durable-threads/scripts/pro5.py
#!/usr/bin/env python3
"""Local config audit and no-generation Codex metadata probe. Never writes user config."""
from __future__ import annotations

import argparse
import asyncio
import json
import math
import re
import subprocess
import sys
import tempfile
import time
from pathlib import Path
from typing import Any

RATE_DATE = "2026-09-22"
PROFILE = '''# Merge these named profiles into config.toml; do not replace existing configuration.
# Verify model availability in your own Codex session before use.
[profiles.dt-pro5]
model = "gpt-6-astra"
model_reasoning_effort = "medium"
service_tier = "default"

[profiles.dt-pro5-deep]
model = "gpt-6-astra"
model_reasoning_effort = "high"
service_tier = "default"

[profiles.dt-pro5-routine]
model = "gpt-5.6-luna"
model_reasoning_effort = "xhigh"
service_tier = "default"
'''


def safe_identifier(value: Any) -> str | None:
    if isinstance(value, str) and re.fullmatch(r"[A-Za-z0-9._-]{1,96}", value):
        return value
    return None


def audit_config(config: dict[str, Any], profile: str | None = None) -> dict[str, Any]:
    """Inspect one config file only, not the resolved managed/project/CLI configuration."""
    if not isinstance(config, dict):
        raise ValueError("configuration must be a table")
    selected = profile or config.get("profile")
    if selected is not None and not isinstance(selected, str):
        raise ValueError("profile must be a string")
    effective = dict(config)
    if selected:
        profiles = config.get("profiles", {})
        if not isinstance(profiles, dict):
            raise ValueError("profiles must be a table")
        layer = profiles.get(selected)
        if not isinstance(layer, dict):
            raise ValueError("selected profile is absent from this configuration")
        for key, value in layer.items():
            if isinstance(value, dict) and isinstance(effective.get(key), dict):
                effective[key] = {**effective[key], **value}
            else:
                effective[key] = value
    findings: list[dict[str, str]] = []

    def flag(code: str, message: str) -> None:
        findings.append({"code": code, "message": message})

    effort = effective.get("model_reasoning_effort")
    if isinstance(effort, str) and effort in {"high", "xhigh", "max", "ultra"}:
        flag(
            "EXPENSIVE_BASELINE",
            "Benchmark Astra medium for routine work; reserve higher effort for hard "
            "decisions. Keep quality checks.",
        )
    tier = effective.get("service_tier")
    if isinstance(tier, str) and tier in {"fast", "priority"}:
        flag(
            "FAST_TIER",
            "Prefer standard speed for allowance efficiency. Published Astra Fast "
            "credit consumption is 2.5x standard.",
        )
    if any(key in effective for key in ("model_context_window", "model_auto_compact_token_limit")):
        flag(
            "MANUAL_CONTEXT",
            "Manual context/compaction overrides need workload evidence. A larger "
            "window is capacity, not free context.",
        )
    agents = effective.get("agents", {})
    if not isinstance(agents, dict):
        raise ValueError("agents must be a table")
    if agents.get("enabled", True) and not agents.get("default_subagent_model"):
        flag(
            "CHILD_INHERITANCE",
            "Unspecified child models may inherit the parent. Select an economical "
            "worker explicitly when delegation is justified.",
        )
    concurrency = agents.get("max_concurrent_threads_per_session", agents.get("max_threads"))
    if isinstance(concurrency, (int, float)) and concurrency > 2:
        flag(
            "FANOUT",
            "A high per-session child limit can amplify spend. Prefer one writer and "
            "an on-demand independent reviewer.",
        )
    servers = effective.get("mcp_servers", {})
    if not isinstance(servers, dict):
        raise ValueError("mcp_servers must be a table")
    count = sum(
        isinstance(item, dict) and item.get("enabled", True) is not False
        for item in servers.values()
    )
    if count:
        flag(
            "MCP_CONTEXT",
            "Enable only tools relevant to the task. Audit schemas and outputs; do "
            "not disable required safety tooling.",
        )
    features = effective.get("features", {})
    experimental = ("reasoning_effort_override", "step_model_switching")
    if isinstance(features, dict) and any(features.get(key) for key in experimental):
        flag(
            "EXPERIMENTAL",
            "Verify the installed server, enabled flags, model support, and "
            "compaction compatibility before live step changes.",
        )
    return {
        "scope": "one-config-file-not-effective-runtime", "profile": safe_identifier(selected),
        "model": safe_identifier(effective.get("model")), "effort": safe_identifier(effort),
        "serviceTier": safe_identifier(effective.get("service_tier")), "enabledMcpServers": count,
        "findings": findings, "writesPerformed": False, "subscriptionMultiplierVerified": False,
    }


def quota_summary(payload: dict[str, Any], *, now: float) -> dict[str, Any]:
    """Expose only the relevant bucket, not emails, account IDs or raw account data."""
    if not isinstance(payload, dict) or not math.isfinite(now) or now < 0:
        raise ValueError("invalid quota observation")
    buckets = payload.get("rateLimitsByLimitId")
    bucket = buckets.get("codex") if isinstance(buckets, dict) else None
    if bucket is None:
        bucket = payload.get("rateLimits")
    if not isinstance(bucket, dict) or bucket.get("limitId") not in (None, "codex"):
        return {"status": "unknown", "windows": []}
    windows = []
    for key in ("primary", "secondary"):
        item = bucket.get(key)
        if item is None:
            continue
        if not isinstance(item, dict):
            return {"status": "unknown", "windows": []}
        used, reset, duration = (
            item.get(name) for name in ("usedPercent", "resetsAt", "windowDurationMins")
        )
        values = (used, reset, duration)
        if any(
            isinstance(v, bool) or not isinstance(v, (int, float)) or not math.isfinite(v)
            for v in values
        ):
            return {"status": "unknown", "windows": []}
        if not 0 <= used <= 100 or reset <= now or duration <= 0:
            return {"status": "refresh_required", "windows": []}
        windows.append({"window": key, "usedPercent": used, "remainingPercent": 100 - used,
                        "resetsAt": reset, "windowDurationMins": duration})
    remaining = min((item["remainingPercent"] for item in windows), default=None)
    reached = bucket.get("rateLimitReachedType") is not None
    status = "observed" if windows else "unknown"
    if reached or (remaining is not None and remaining <= 15):
        status = "pause"
    return {"status": status,
            "windows": windows, "remainingPercent": remaining, "observedAt": now,
            "reservePercent": 15, "reserveIsLocalPolicy": True}


async def probe_metadata(command: list[str], timeout: float = 20) -> dict[str, Any]:
    """Uses only initialize/account-read/model-list/rate-limit-read; never starts a turn."""
    process = await asyncio.create_subprocess_exec(
        *command, stdin=asyncio.subprocess.PIPE, stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.DEVNULL, limit=1_048_576,
    )
    next_id = 0
    methods: list[str] = []

    async def send(message: dict[str, Any]) -> None:
        if process.stdin is None:
            raise RuntimeError("Codex input unavailable")
        process.stdin.write((json.dumps(message) + "\n").encode())
        await process.stdin.drain()

    async def rpc(method: str, params: dict[str, Any] | None = None) -> dict[str, Any]:
        nonlocal next_id
        next_id += 1
        identifier = next_id
        methods.append(method)
        message: dict[str, Any] = {"id": identifier, "method": method}
        if params is not None:
            message["params"] = params
        await send(message)
        for _ in range(1000):
            if process.stdout is None:
                raise RuntimeError("Codex output unavailable")
            raw = await process.stdout.readline()
            if not raw:
                raise RuntimeError("Codex exited before answering metadata request")
            response = json.loads(raw)
            if not isinstance(response, dict):
                raise ValueError("invalid Codex metadata response")
            if "method" in response and "id" in response:
                # Deny every server-initiated action. The probe is never an approver.
                await send({
                    "id": response["id"],
                    "error": {"code": -32601, "message": "Read-only probe"},
                })
            if response.get("id") == identifier and "method" not in response:
                if "error" in response:
                    raise RuntimeError(f"Codex rejected {method}")
                result = response.get("result")
                if not isinstance(result, dict):
                    raise ValueError("invalid Codex result")
                return result
        raise RuntimeError("Codex metadata notification limit exceeded")

    async def run() -> dict[str, Any]:
        await rpc("initialize", {
            "clientInfo": {"name": "durable_threads_probe", "version": "0.6.0"},
            "capabilities": {"experimentalApi": False},
        })
        await send({"method": "initialized"})
        account_payload = await rpc("account/read", {"refreshToken": False})
        account = account_payload.get("account")
        account_type = safe_identifier(account.get("type")) if isinstance(account, dict) else None
        # Do not output the raw account object: it can include email and account identifiers.
        result: dict[str, Any] = {
            "authType": account_type, "chatgptSignIn": account_type == "chatgpt",
            "subscriptionMultiplierVerified": False, "models": [], "warnings": [],
        }
        cursor = None
        for page in range(8):
            params: dict[str, Any] = {"limit": 100}
            if cursor:
                params["cursor"] = cursor
            model_page = await rpc("model/list", params)
            for model in model_page.get("data", []):
                if not isinstance(model, dict):
                    continue
                identifier = safe_identifier(model.get("model") or model.get("id"))
                if identifier not in {"gpt-6-astra", "gpt-5.6-luna", "gpt-5.6-terra"}:
                    continue
                efforts = model.get("supportedReasoningEfforts", [])
                result["models"].append({"model": identifier, "supportedEfforts": [
                    safe_identifier(item.get("reasoningEffort") or item.get("effort"))
                    for item in efforts if isinstance(item, dict)
                ]})
            cursor = model_page.get("nextCursor")
            if not cursor:
                break
            if page == 7:
                result["warnings"].append("Model listing truncated after eight pages")
        if account_type == "chatgpt":
            try:
                limits = await rpc("account/rateLimits/read")
                result["quota"] = quota_summary(limits, now=time.time())
            except RuntimeError:
                result["quota"] = {"status": "unknown", "windows": []}
                result["warnings"].append("Rate-limit read unavailable; check /status")
        else:
            result["warnings"].append(
                "ChatGPT-managed sign-in not verified; do not assume subscription billing"
            )
        result.update({"requestedMethods": methods, "modelGenerationsRequested": 0,
                       "stepSwitchingVerified": False, "writesToUserConfig": False})
        return result

    try:
        return await asyncio.wait_for(run(), timeout)
    finally:
        if process.returncode is None:
            try:
                process.terminate()
            except ProcessLookupError:
                pass
            try:
                await asyncio.wait_for(process.wait(), 3)
            except asyncio.TimeoutError:
                try:
                    process.kill()
                except ProcessLookupError:
                    pass
                await process.wait()


def schema_probe(codex: str) -> dict[str, Any]:
    """Schema presence is not proof that a live turn or model accepts an update."""
    with tempfile.TemporaryDirectory(prefix="dt-schema-") as directory:
        completed = subprocess.run(
            [codex, "app-server", "generate-json-schema", "--experimental", "--out", directory],
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=20, check=False,
        )
        if completed.returncode != 0:
            return {"status": "unavailable", "liveSwitchingVerified": False}
        found = any('turn/settings/update' in path.read_text(encoding="utf-8")
                    for path in Path(directory).rglob("*.json") if path.stat().st_size < 20_000_000)
        return {"status": "schema_only", "turnSettingsUpdateAdvertised": found,
                "liveSwitchingVerified": False}


def read_toml(path: Path) -> dict[str, Any]:
    try:
        import tomllib
    except ImportError:
        try:
            import tomli as tomllib
        except ImportError as exc:
            raise RuntimeError("Use Python 3.11+ or install tomli for the TOML audit") from exc
    if path.stat().st_size > 2_000_000:
        raise ValueError("configuration is too large")
    with path.open("rb") as stream:
        return tomllib.load(stream)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    commands = parser.add_subparsers(dest="command", required=True)
    audit = commands.add_parser("audit", help="read one config file without displaying secrets")
    audit.add_argument("--config", type=Path, default=Path.home() / ".codex" / "config.toml")
    audit.add_argument("--profile")
    probe = commands.add_parser(
        "probe", help="inspect existing Codex sign-in, models and quota; no turns"
    )
    probe.add_argument("--codex", default="codex")
    probe.add_argument("--timeout", type=float, default=20)
    probe.add_argument(
        "--schema", action="store_true", help="also generate experimental schema temporarily"
    )
    schema = commands.add_parser("schema", help="inspect schema without starting a model turn")
    schema.add_argument("--codex", default="codex")
    commands.add_parser("profile", help="print named TOML profiles without installing them")
    args = parser.parse_args(argv)
    try:
        if args.command == "profile":
            print(PROFILE, end="")
            return 0
        if args.command == "schema":
            result = schema_probe(args.codex)
        elif args.command == "audit":
            result = audit_config(read_toml(args.config.expanduser()), args.profile)
        else:
            if not math.isfinite(args.timeout) or not 1 <= args.timeout <= 120:
                raise ValueError("timeout must be between 1 and 120 seconds")
            result = asyncio.run(probe_metadata([args.codex, "app-server"], args.timeout))
            if args.schema:
                result["experimentalSchema"] = schema_probe(args.codex)
        print(json.dumps(result, indent=2, sort_keys=True))
        return 0
    except (
        OSError, RuntimeError, ValueError, asyncio.TimeoutError, subprocess.TimeoutExpired
    ) as exc:
        # TOML error text and filesystem exceptions can contain private values/paths.
        failure = {
            "error": type(exc).__name__,
            "message": (
                "Audit/probe failed; check file syntax, selected profile, executable, "
                "and timeout locally."
            ),
        }
        print(json.dumps(failure), file=sys.stderr)
        return 2


if __name__ == "__main__":
    raise SystemExit(main())
